<a href="https://colab.research.google.com/github/feyza-inl/Su-Alti-Goruntu-Iyilestirme/blob/main/REAL_ESRGAN4ipynb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install basicsr==1.4.2 --no-deps
!pip install realesrgan==0.3.0

dosya = '/usr/local/lib/python3.12/dist-packages/basicsr/data/degradations.py'
with open(dosya, 'r') as f:
    icerik = f.read()
icerik = icerik.replace(
    'from torchvision.transforms.functional_tensor import rgb_to_grayscale',
    'from torchvision.transforms.functional import rgb_to_grayscale'
)
with open(dosya, 'w') as f:
    f.write(icerik)
print('Kurulum tamam!')

Kurulum tamam!


In [ ]:
!wget https://github.com/xinntao/Real-ESRGAN/releases/download/v0.1.0/RealESRGAN_x4plus.pth -P /content/weights/

from realesrgan import RealESRGANer
from basicsr.archs.rrdbnet_arch import RRDBNet

model = RRDBNet(num_in_ch=3, num_out_ch=3, num_feat=64, num_block=23, num_grow_ch=32, scale=4)
upsampler = RealESRGANer(
    scale=4,
    model_path='/content/weights/RealESRGAN_x4plus.pth',
    model=model,
    tile=400,
    tile_pad=10,
    pre_pad=0,
    half=True
)
print('Model hazır!')

--2026-03-04 22:09:31--  https://github.com/xinntao/Real-ESRGAN/releases/download/v0.1.0/RealESRGAN_x4plus.pth
Resolving github.com (github.com)... 140.82.114.4
Connecting to github.com (github.com)|140.82.114.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://release-assets.githubusercontent.com/github-production-release-asset/387326890/08f0e941-ebb7-48f0-9d6a-73e87b710e7e?sp=r&sv=2018-11-09&sr=b&spr=https&se=2026-03-04T23%3A01%3A22Z&rscd=attachment%3B+filename%3DRealESRGAN_x4plus.pth&rsct=application%2Foctet-stream&skoid=96c2d410-5711-43a1-aedd-ab1947aa7ab0&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skt=2026-03-04T22%3A00%3A28Z&ske=2026-03-04T23%3A01%3A22Z&sks=b&skv=2018-11-09&sig=SQeakN1GknGkGsmHD4z8JkEpUnSquPG4juWTAJlIGto%3D&jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmVsZWFzZS1hc3NldHMuZ2l0aHVidXNlcmNvbnRlbnQuY29tIiwia2V5Ijoia2V5MSIsImV4cCI6MTc3MjY2MzkyOCwibmJmIjoxNzcyNjYyMTI4LCJwYXRoIjoicmVsZWFzZWFzc2V0cHJvZH

In [ ]:
import cv2
import os

raw_klasor = '/content/drive/MyDrive/raw-890'
iyilestirilmis_klasor = '/content/drive/MyDrive/iyilestirilmis-890-v2'
os.makedirs(iyilestirilmis_klasor, exist_ok=True)

dosyalar = os.listdir(raw_klasor)
toplam = len(dosyalar)

for i, dosya in enumerate(dosyalar):
    raw_yol = os.path.join(raw_klasor, dosya)
    cikti_yol = os.path.join(iyilestirilmis_klasor, dosya)

    if os.path.exists(cikti_yol):
        continue

    img = cv2.imread(raw_yol, cv2.IMREAD_COLOR)
    if img is None:
        continue

    output, _ = upsampler.enhance(img, outscale=4)
    cv2.imwrite(cikti_yol, output)
    print(f'{i+1}/{toplam} → {dosya}')

print('Tamamlandı!')

	Tile 1/12
	Tile 2/12
	Tile 3/12
	Tile 4/12
	Tile 5/12
	Tile 6/12
	Tile 7/12
	Tile 8/12
	Tile 9/12
	Tile 10/12
	Tile 11/12
	Tile 12/12
216/890 → 279_img_.png
	Tile 1/4
	Tile 2/4
	Tile 3/4
	Tile 4/4
217/890 → 27_img_.png
	Tile 1/4
	Tile 2/4
	Tile 3/4
	Tile 4/4
218/890 → 280_img_.png
	Tile 1/4
	Tile 2/4
	Tile 3/4
	Tile 4/4
219/890 → 281_img_.png
	Tile 1/1
220/890 → 282_img_.png
	Tile 1/1
221/890 → 284_img_.png
	Tile 1/4
	Tile 2/4
	Tile 3/4
	Tile 4/4
222/890 → 285_img_.png
	Tile 1/12
	Tile 2/12
	Tile 3/12
	Tile 4/12
	Tile 5/12
	Tile 6/12
	Tile 7/12
	Tile 8/12
	Tile 9/12
	Tile 10/12
	Tile 11/12
	Tile 12/12
223/890 → 287_img_.png
	Tile 1/8
	Tile 2/8
	Tile 3/8
	Tile 4/8
	Tile 5/8
	Tile 6/8
	Tile 7/8
	Tile 8/8
224/890 → 2882.png
	Tile 1/1
225/890 → 288_img_.png
	Tile 1/9
	Tile 2/9
	Tile 3/9
	Tile 4/9
	Tile 5/9
	Tile 6/9
	Tile 7/9
	Tile 8/9
	Tile 9/9
226/890 → 28_img_.png
	Tile 1/1
227/890 → 290_img_.png
	Tile 1/1
228/890 → 292_img_.png
	Tile 1/1
229/890 → 293_img_.png
	Tile 1/1
230/890 → 294_

KeyboardInterrupt: 

In [ ]:
import numpy as np
from skimage.metrics import peak_signal_noise_ratio as psnr
from skimage.metrics import structural_similarity as ssim

raw_klasor = '/content/drive/MyDrive/raw-890'
ref_klasor = '/content/drive/MyDrive/reference-890-extracted/reference-890'
iyi_klasor = '/content/drive/MyDrive/iyilestirilmis-890-v2'

psnr_raw_listesi = []
ssim_raw_listesi = []
psnr_iyi_listesi = []
ssim_iyi_listesi = []

for dosya in os.listdir(raw_klasor):
    raw_yol = os.path.join(raw_klasor, dosya)
    ref_yol = os.path.join(ref_klasor, dosya)
    iyi_yol = os.path.join(iyi_klasor, dosya)

    if not os.path.exists(ref_yol) or not os.path.exists(iyi_yol):
        continue

    raw = cv2.imread(raw_yol)
    ref = cv2.imread(ref_yol)
    iyi = cv2.imread(iyi_yol)

    if raw is None or ref is None or iyi is None:
        continue

    hedef = (raw.shape[1], raw.shape[0])
    ref_r = cv2.resize(ref, hedef)
    iyi_r = cv2.resize(iyi, hedef)

    psnr_raw_listesi.append(psnr(ref_r, raw))
    ssim_raw_listesi.append(ssim(ref_r, raw, channel_axis=2))
    psnr_iyi_listesi.append(psnr(ref_r, iyi_r))
    ssim_iyi_listesi.append(ssim(ref_r, iyi_r, channel_axis=2))

print(f'Karşılaştırılan görüntü: {len(psnr_raw_listesi)}')
print()
print('========== SONUÇLAR ==========')
print(f'HAM GÖRÜNTÜ     → PSNR: {np.mean(psnr_raw_listesi):.2f} dB | SSIM: {np.mean(ssim_raw_listesi):.4f}')
print(f'İYİLEŞTİRİLMİŞ  → PSNR: {np.mean(psnr_iyi_listesi):.2f} dB | SSIM: {np.mean(ssim_iyi_listesi):.4f}')
fark_psnr = np.mean(psnr_iyi_listesi) - np.mean(psnr_raw_listesi)
fark_ssim = np.mean(ssim_iyi_listesi) - np.mean(ssim_raw_listesi)
print(f'FARK            → PSNR: {fark_psnr:+.2f} dB | SSIM: {fark_ssim:+.4f}')

Karşılaştırılan görüntü: 296

========== SONUÇLAR ==========
HAM GÖRÜNTÜ     → PSNR: 17.06 dB | SSIM: 0.7463
İYİLEŞTİRİLMİŞ  → PSNR: 16.63 dB | SSIM: 0.6642
FARK            → PSNR: -0.43 dB | SSIM: -0.0820


In [ ]:
import cv2
import os

raw_klasor = '/content/drive/MyDrive/raw-890'
iyi_klasor = '/content/drive/MyDrive/iyilestirilmis-890-v2'

dosyalar = list(os.listdir(iyi_klasor))[:5]

for dosya in dosyalar:
    raw = cv2.imread(os.path.join(raw_klasor, dosya))
    iyi = cv2.imread(os.path.join(iyi_klasor, dosya))

    if raw is None or iyi is None:
        continue

    buyutme = iyi.shape[0] / raw.shape[0]
    print(f'{dosya}:')
    print(f'  RAW: {raw.shape[1]}x{raw.shape[0]}')
    print(f'  İYİ: {iyi.shape[1]}x{iyi.shape[0]}')
    print(f'  Büyütme: {buyutme:.1f}x')
    print('---')

186_img_.png:
  RAW: 1200x771
  İYİ: 4800x3084
  Büyütme: 4.0x
---
160_img_.png:
  RAW: 800x1150
  İYİ: 3200x4600
  Büyütme: 4.0x
---
15704.png:
  RAW: 1280x720
  İYİ: 5120x2880
  Büyütme: 4.0x
---
192_img_.png:
  RAW: 640x480
  İYİ: 2560x1920
  Büyütme: 4.0x
---
157_img_.png:
  RAW: 640x480
  İYİ: 2560x1920
  Büyütme: 4.0x
---


PSNR yorumu: 17.06 → 16.63, düşüş 2x'e göre daha fazla (-0.43 vs -0.30). 4x büyütünce model daha fazla değişiklik yapıyor, referanstan daha çok uzaklaşıyor.
SSIM yorumu: 0.7463 → 0.6642, yine ciddi düşüş ama 2x'e göre biraz daha iyi (-0.0820 vs -0.0850). 4x daha fazla detay ekliyor ama yapısal benzerlik yine kaybolmuş.
Genel: 4x daha agresif iyileştirme yapıyor. Göze güzel görünüyor ama UIEB referansının beklediği sualtı renk düzeltmesini yapmadığı için metrikler yine düşüyor.

In [ ]:
import cv2
import numpy as np
import os

def uiqm(img):
    img = img.astype(np.float32) / 255.0
    r, g, b = img[:,:,2], img[:,:,1], img[:,:,0]

    rg = r - g
    yb = (r + g) / 2 - b
    uicm = -0.0268 * np.sqrt(np.mean(rg**2) + np.mean(yb**2)) + 0.1586

    def sobel(ch):
        kx = np.array([[-1,0,1],[-2,0,2],[-1,0,1]], np.float32)
        ky = kx.T
        gx = cv2.filter2D(ch, -1, kx)
        gy = cv2.filter2D(ch, -1, ky)
        return np.abs(gx) + np.abs(gy)

    uism = 0.299 * np.mean(sobel(r)) + 0.587 * np.mean(sobel(g)) + 0.114 * np.mean(sobel(b))
    uiconm = np.log(np.std(img) + 1e-5)

    return 0.282 * uicm + 0.287 * uism + 0.433 * uiconm

def uciqe(img):
    # 0-1 aralığına normalize et
    img_norm = img.astype(np.float32) / 255.0
    img_lab = cv2.cvtColor(img_norm, cv2.COLOR_BGR2Lab)

    l = img_lab[:,:,0] / 100.0      # 0-1 arası
    a = img_lab[:,:,1] / 127.0      # normalize
    b = img_lab[:,:,2] / 127.0      # normalize

    chroma = np.sqrt(a**2 + b**2)
    sigma_c = np.std(chroma)

    mu_l = np.mean(l)

    saturation = chroma / (np.mean(chroma) + 1e-5)
    mu_s = np.std(saturation)

    return 0.4680 * sigma_c + 0.2745 * mu_l + 0.2576 * mu_s

# --- Ana kod ---
raw_klasor = '/content/drive/MyDrive/raw-890'
iyi_klasor = '/content/drive/MyDrive/iyilestirilmis-890-v2'

uiqm_raw = []
uiqm_iyi = []
uciqe_raw = []
uciqe_iyi = []

dosyalar = os.listdir(iyi_klasor)
toplam = len(dosyalar)

for i, dosya in enumerate(dosyalar):
    raw_yol = os.path.join(raw_klasor, dosya)
    iyi_yol = os.path.join(iyi_klasor, dosya)

    if not os.path.exists(raw_yol):
        continue

    raw = cv2.imread(raw_yol)
    iyi = cv2.imread(iyi_yol)

    if raw is None or iyi is None:
        continue

    iyi_r = cv2.resize(iyi, (raw.shape[1], raw.shape[0]))

    uiqm_raw.append(uiqm(raw))
    uiqm_iyi.append(uiqm(iyi_r))
    uciqe_raw.append(uciqe(raw))
    uciqe_iyi.append(uciqe(iyi_r))

    if (i+1) % 50 == 0:
        print(f'{i+1}/{toplam} işlendi...')

print()
print('========== SONUÇLAR ==========')
print(f'HAM GÖRÜNTÜ     → UIQM: {np.mean(uiqm_raw):.4f} | UCIQE: {np.mean(uciqe_raw):.4f}')
print(f'İYİLEŞTİRİLMİŞ  → UIQM: {np.mean(uiqm_iyi):.4f} | UCIQE: {np.mean(uciqe_iyi):.4f}')
fark_uiqm = np.mean(uiqm_iyi) - np.mean(uiqm_raw)
fark_uciqe = np.mean(uciqe_iyi) - np.mean(uciqe_raw)
print(f'FARK            → UIQM: {fark_uiqm:+.4f} | UCIQE: {fark_uciqe:+.4f}')

50/296 işlendi...
100/296 işlendi...
150/296 işlendi...
200/296 işlendi...
250/296 işlendi...

========== SONUÇLAR ==========
HAM GÖRÜNTÜ     → UIQM: -0.6446 | UCIQE: 0.2647
İYİLEŞTİRİLMİŞ  → UIQM: -0.6517 | UCIQE: 0.2710
FARK            → UIQM: -0.0071 | UCIQE: +0.0063


UCIQE:Pozitif! Real-ESRGAN renk canlılığını, parlaklığı ve kontrastı biraz artırdı. Görsel kalite iyileşti.
UIQM:Negatif ve zaten çok düşük. İki nedeni var:
1 — UIQM sualtı görüntüleri için tasarlanmış ama renk dengesi bozuk görüntülerde negatif çıkabiliyor.
2 — Real-ESRGAN keskinleştirme yaparken sualtı renk dengesini daha da bozuyor olabilir.